PWGMM/UE/Tasks/ue-zdc-analysys.cxx

```diff
@@ -179,41 +188,68 @@ struct ZDCAnalysis {
   PROCESS_SWITCH(ZDCAnalysis, processZdcBcAss, "Processing ZDC w. BC association", true);
 
   void processZdcCollAss(
-    soa::Join<aod::Collisions, aod::EvSels>::iterator const& collision,
+    ColEvSels const& cols,
+    BCsRun3 const& bcs,
     aod::Zdcs const& zdcs)
   {
-    if (collision.foundZDCId() >= 0) {
-      registry.get<TH1>(HIST("ZNAcoll"))->Fill(collision.foundZDC().amplitudeZNA());
-      registry.get<TH1>(HIST("ZNCcoll"))->Fill(collision.foundZDC().amplitudeZNC());
-      registry.get<TH1>(HIST("ZPAcoll"))->Fill(collision.foundZDC().amplitudeZPA());
-      registry.get<TH1>(HIST("ZPCcoll"))->Fill(collision.foundZDC().amplitudeZPC());
-      registry.get<TH1>(HIST("ZEM1coll"))->Fill(collision.foundZDC().amplitudeZEM1());
-      registry.get<TH1>(HIST("ZEM2coll"))->Fill(collision.foundZDC().amplitudeZEM2());
-      registry.get<TH2>(HIST("ZNvsZEMcoll"))->Fill(collision.foundZDC().amplitudeZEM1() + collision.foundZDC().amplitudeZEM2(), collision.foundZDC().amplitudeZNA() + collision.foundZDC().amplitudeZNC());
-      registry.get<TH2>(HIST("ZNAvsZNCcoll"))->Fill(collision.foundZDC().amplitudeZNC(), collision.foundZDC().amplitudeZNA());
-      registry.get<TH2>(HIST("ZPAvsZPCcoll"))->Fill(collision.foundZDC().amplitudeZPC(), collision.foundZDC().amplitudeZPA());
-      registry.get<TH2>(HIST("ZNAvsZPAcoll"))->Fill(collision.foundZDC().amplitudeZPA(), collision.foundZDC().amplitudeZNA());
-      registry.get<TH2>(HIST("ZNCvsZPCcoll"))->Fill(collision.foundZDC().amplitudeZPC(), collision.foundZDC().amplitudeZNC());
-      //
-      registry.get<TH2>(HIST("ZNCadcvstdccoll"))->Fill(collision.foundZDC().timeZNC(), collision.foundZDC().amplitudeZNC());
-      registry.get<TH2>(HIST("ZNAadcvstdccoll"))->Fill(collision.foundZDC().timeZNA(), collision.foundZDC().amplitudeZNA());
-      registry.get<TH2>(HIST("ZPCadcvstdccoll"))->Fill(collision.foundZDC().timeZPC(), collision.foundZDC().amplitudeZPC());
-      registry.get<TH2>(HIST("ZPAadcvstdccoll"))->Fill(collision.foundZDC().timeZPA(), collision.foundZDC().amplitudeZPA());
-      registry.get<TH2>(HIST("ZEM1adcvstdccoll"))->Fill(collision.foundZDC().timeZEM1(), collision.foundZDC().amplitudeZEM1());
-      registry.get<TH2>(HIST("ZEM2adcvstdccoll"))->Fill(collision.foundZDC().timeZEM2(), collision.foundZDC().amplitudeZEM2());
+    // collision-based event selection
+    for (auto& collision : cols) {
+      const auto& foundBC = collision.foundBC_as<BCsRun3>();
+      if (foundBC.has_zdc()) {
+        if (TDCcut) {
+          if ((foundBC.zdc().timeZNA() >= tdcZNmincut) && (foundBC.zdc().timeZNA() <= tdcZNmaxcut))
```

Comment:
I recommend to store `foundBC.zdc()` in a local variable, `auto zdc = foundBC.zdc();`, so that it can be re-used. Creating temporary iterator copy on each call to `.zdc()` is quite inefficient.